homeassistant/components/stream/worker.py

```diff
@@ -60,32 +43,133 @@ def stream_worker(hass, stream, quit_event):
     except (KeyError, IndexError):
         _LOGGER.error("Stream has no video")
         return
+    try:
+        audio_stream = container.streams.audio[0]
+    except (KeyError, IndexError):
+        audio_stream = None
 
-    audio_frame = generate_audio_frame()
-
-    first_packet = True
-    # Holds the buffers for each stream provider
-    outputs = {}
-    # Keep track of the number of segments we've processed
-    sequence = 1
-    # Holds the generated silence that needs to be muxed into the output
-    audio_packets = {}
-    # The presentation timestamp of the first video packet we receive
-    first_pts = 0
-    # The decoder timestamp of the latest packet we processed
+    # The presentation timestamps of the first packet in each stream we receive
+    # Use to adjust before muxing or outputting, but we don't adjust internally
+    first_pts = {}
+    # The decoder timestamps of the latest packet in each stream we processed
     last_dts = None
     # Keep track of consecutive packets without a dts to detect end of stream.
     last_packet_was_without_dts = False
-    # The pts at the beginning of the segment
-    segment_start_v_pts = 0
-    segment_start_a_pts = 0
+    # Holds the buffers for each stream provider
+    outputs = None
+    # Keep track of the number of segments we've processed
+    sequence = 0
+    # The video pts at the beginning of the segment
+    segment_start_pts = None
+    # Because of problems 1 and 2 below, we need to store the first few packets and replay them
+    initial_packets = deque()
+
+    # Have to work around two problems with RTSP feeds in ffmpeg
+    # 1 - first frame has bad pts/dts https://trac.ffmpeg.org/ticket/5018
+    # 2 - seeking can be problematic https://trac.ffmpeg.org/ticket/7815
+
+    def peek_first_pts():
+        nonlocal first_pts, audio_stream
+
+        def empty_stream_dict():
+            return {
+                video_stream: None,
+                **({audio_stream: None} if audio_stream else {}),
+            }
+
+        try:
+            first_packet = empty_stream_dict()
+            first_pts = empty_stream_dict()
+            # Get to first video keyframe
+            while first_packet[video_stream] is None:
+                packet = next(container.demux())
+                if packet.stream == video_stream and packet.is_keyframe:
+                    first_packet[video_stream] = packet
+                    initial_packets.append(packet)
+            # Get first_pts from subsequent frame to first keyframe
+            while any(
+                [pts is None for pts in {**first_packet, **first_pts}.values()]
+            ) and (len(initial_packets) < PACKETS_TO_WAIT_FOR_AUDIO):
+                packet = next(container.demux((video_stream, audio_stream)))
```

Comment:
Had to add this so PyAV only demuxes AV packets ignoring data and subtitles. Not sure of any IP cameras that send subtitles, but many will send events on the data stream that shouldn't be processed (and indeed blew up this logic).